teps/0047-finallytask-execution-post-timeout.md

```diff
@@ -0,0 +1,265 @@
+| ∙ hello-world-pipeline-run-with-timeout-task1-wgcq7 | task1     | 38 seconds ago | 30 seconds | Failed(TaskRunTimeout) |
+|                                                     |           |                |            |                        |
+
+
+The finally task still executes after the task failure.
+
+
+Finally if we reduce the pipelinerun timeout to 10s, our status report shows:
+
+`PipelineRun "hello-world-pipeline-run-with-timeout" failed to finish within "10s" (TaskRun "hello-world-pipeline-run-with-timeout-task1-q7fw4" failed to finish within "30s")`
+
+| NAME                                                | TASK NAME | STARTED       | DURATION   | STATUS                 |
+|-----------------------------------------------------|-----------|---------------|------------|------------------------|
+| ∙ hello-world-pipeline-run-with-timeout-task1-q7fw4 | task1     | 2 minutes ago | 30 seconds | Failed(TaskRunTimeout) |
+|                                                     |           |               |            |                        |
+|                                                     |           |               |            |                        |
+
+
+The pipelinerun timeout take precedence over the task timeout. After 10s the task fails... And the finally task does not get the chance to execute.
+
+
+For this reason, it is currently not possible to rely on Finally tasks for any of the aforementioned use cases.
+
+### Goals
+
+<!--
+List the specific goals of the TEP.  What is it trying to achieve?  How will we
+know that this has succeeded?
+-->
+
+Enable the uses cases :
+
+- Cleanup cluster resources after finishing (with success/failure) integration tests (Dogfooding Scenario)
+- Update Pull Request with what happened overall in the pipeline (pipeline level)
+- Report Test Results at the end of the test pipeline (Notifications Scenario)
+
+When a pipelinerun times out.
+
+### Non-Goals
+
+<!--
+What is out of scope for this TEP?  Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation.  The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+Enable finally task to run when a pipeline times out. This implies a behavioral change, as finally tasks will run no matter what. 
+
+Enable pipeline authors to specify a timeout field for finally tasks. In all normal run, that timeout is not needed and finally tasks execute after non-finally tasks. But in case of timed out pipeline, the finally task execution is bounded by the declared timeout.
+
+```yaml
+spec:
+  tasks:
+    - name: tests
+      taskRef:
+        Name: integration-test
+  finally:
+    timeout: "0h0m10s"
```

Comment:
Please confirm, are you proposing a new field `timeout` under `finally` section which will apply to all finally tasks defined? Also what if each finally task has specified its own timeout plus this new field?

Reply:
yes exactly. It makes sense to me as all finally task are run in parallel. If a task has a timeout declared, the behavior is unchanged for the task, but the finally timeout field will take precedence.

If a task has a 20s timeout and the finally timeout field is 30s. Then the task timeout is relevant. The other way around, with a 30s task timeout and a 20s finally timeout, the task fails after 20s.

Reply:
hmm this would be a backwards incompatible change, and would be very difficult as shown here. currently `finally` is a list, this example shows it as a hybrid of a dictionary and a list. if you wanted to have a `timeout` section in `finally` you'd probably need to add a tasks section or similar as well, e.g.:

```
finally:
  timeout: "0h0m10s"
  tasks:
  - ...
```

This would be a backwards incompatible change tho, so we'd need to support both for a while (@jerop was dealing with a similar issue for adding more configuration to when expressions - this feels like a lesson about the consequences of using lists).

Reply:
yup this would be a backward incompatible change, instead how about adding a new field under `spec`:

```
kind: Pipeline
spec:
  finallyTimeout: ...
  tasks:
      ....
   finally:
      ...
```

or introduce a new field in `pipelineRun`.

Reply:
we have two different options to specify timeout today (1) at the `pipelineTask` level (2) at the `pipelineRun` level. 

Timeout at the `pipelineRun` level:

```
apiVersion: tekton.dev/v1beta1
kind: PipelineRun
metadata:
  name: hello-world-pipeline-run-with-timeout
spec:
  timeout: "0h0m60s"
```

Finally tasks are bound to this time out (or default `pipelineRun` timeout of 60 minutes). We are trying to override the `pipelineRun` `timeout` in this proposal. And, I recommend to add one more field at the same `pipelineRun` level which overrides `pipelineRun` timeout and the finally tasks would honor this new field:

```
apiVersion: tekton.dev/v1beta1
kind: PipelineRun
metadata:
  name: hello-world-pipeline-run-with-timeout
spec:
  timeout: "0h0m60s"
  finallyTimeout: "0h0m60s"
```

Reply:
thanks @bobcatfish @pritidesai for the review.

I am wondering if adding a second timeout field at the `pipelinerun` spec level is the best strategy. As a user, it might be confusing, could make more difficult to understand the behavior.

What if we step back here and instead re-use the timeout field in the `pipelineTask`.  We would still override the `pipelinerun` timeout, but apply it for each finally task.